### PR TITLE
CE-2531 Restore certain revisions of MediaWiki JS files using auto-approving mechanism

### PR DESCRIPTION
**Pull request by adamkarminski**
Hey @Wikia/community-engineering !

This PR is focused on two major changes:
1. Providing a way for reviewers to auto-approve their revisions.
2. Creating a Special:ContentReview subpages mechanism with archives of reviews for wikias. A single archive can be found at `http://dev.wikia.com/Special:ContentReview/{wikiId}`.

It also introduces a new class - `ContentReviewService` which goal is to be the main controlling level of the ContentReview tool soon making the `ContentApiController` just a wrapper callable via http and available publicly.

Have fun!

/cc @kamilkoterba @Grunny

### CODE REVIEW

File: extensions/wikia/ContentReview/ContentReviewHelper.php [hunk 1]
```diff
@@ -354,9 +354,25 @@ public function replaceWithLastApproved( \Title $title, $contentType, &$text ) {
...
 	}
 
+	/**
+	 * Checks if a user can edit a JS page in the MediaWiki namespace.
+	 * @param \Title $title
+	 * @param \User $user
+	 * @return bool
+	 */
 	public function userCanEditJsPage( \Title $title, \User $user ) {
 		return $title->inNamespace( NS_MEDIAWIKI )
```
Comment (kamilkoterba) [inline extensions/wikia/ContentReview/ContentReviewHelper.php]: $title->inNamespace( NS_MEDIAWIKI )
redundant check, right?

Comment (adamkarminski) [inline extensions/wikia/ContentReview/ContentReviewHelper.php]: Why?

File: extensions/wikia/ContentReview/ContentReviewHelper.php [hunk 2]
```diff
@@ -354,9 +354,25 @@ public function replaceWithLastApproved( \Title $title, $contentType, &$text ) {
...
 	}
+
+	/**
+	 * Checks if a user is a reviewer entitled to an automatic approval and if he requested it.
+	 * @param \User $user
+	 * @return bool
+	 */
+	public function userCanAutomaticallyApprove( \User $user ) {
+		return $user->isAllowed( 'content-review' )
+			&& $user->getRequest()->getBool( 'wpApprove' );
```
Comment (kamilkoterba) [inline extensions/wikia/ContentReview/ContentReviewHelper.php:380]: Can we add explanation that wp prefix refers to WikiaPage and wpApprove is a checkbox near publish (wpSave) button on edit page?

Maybe it's not very universal description but explains naming pick and place of use. Might save some time of searching the code next time.

File: extensions/wikia/ContentReview/ContentReviewHelper.php (hunk 1 above)
Comment (kamilkoterba) [inline extensions/wikia/ContentReview/ContentReviewHelper.php]: isJsPage does it 
https://github.com/Wikia/app/blob/dev/includes/Title.php#L1110-1110

Comment (adamkarminski) [inline extensions/wikia/ContentReview/ContentReviewHelper.php]: :+1:

File: extensions/wikia/ContentReview/ContentReviewHelper.php (hunk 2 above)
Comment (adamkarminski) [inline extensions/wikia/ContentReview/ContentReviewHelper.php:380]: Yes, we can.

File: extensions/wikia/ContentReview/controllers/ContentReviewSpecialController.class.php [hunk 3]
```diff
@@ -39,9 +46,22 @@ public function index() {
...
-		$model = new ReviewModel();
-		$reviews = $model->getContentToReviewFromDatabase();
-		$this->reviews = $this->prepareReviewData( $reviews );
+		$wikiId = $this->getPar();
+
+		if ( !empty( $wikiId ) ) {
+			$logModel = new ReviewLogModel();
+			$revisionModel = new CurrentRevisionModel();
+
+			$reviews = $logModel->getArchivedReviewsForWiki( $wikiId );
```
Comment (kamilkoterba) [inline extensions/wikia/ContentReview/controllers/ContentReviewSpecialController.class.php:55]: What's our approach to caching in this feature. Do we have any plans?

Comment (adamkarminski) [inline extensions/wikia/ContentReview/controllers/ContentReviewSpecialController.class.php:55]: Not now, as this tool is available for ~10 people and will be rarely used.

File: extensions/wikia/ContentReview/ContentReviewHelper.php (hunk 2 above)
Comment (adamkarminski) [inline extensions/wikia/ContentReview/ContentReviewHelper.php:380]: Done! See the docs for the method.

File: extensions/wikia/ContentReview/controllers/ContentReviewSpecialController.class.php [hunk 4]
```diff
@@ -84,4 +104,39 @@ private function prepareReviewData( $reviewsRaw ) {
...
 	}
+
+	private function prepareArchivedReviewData( $reviewsRaw, $reviewed ) {
+		$reviews = [];
+
+		foreach ( $reviewsRaw as $review ) {
+			$title = GlobalTitle::newFromID( $review['page_id'], $review['wiki_id'] );
+
+			$review['url'] = $title->getFullURL();
+			$review['title'] = $title->getBaseText();
```
Comment (kamilkoterba) [inline extensions/wikia/ContentReview/controllers/ContentReviewSpecialController.class.php]: Can we use getText to distinguish subpages like Wikia.js/jsSliderNew.js? that are separate js pages.

Comment (adamkarminski) [inline extensions/wikia/ContentReview/controllers/ContentReviewSpecialController.class.php]: Done

File: extensions/wikia/EditPageLayout/templates/EditPageLayout_CodeButtons.php [hunk 5]
```diff
@@ -1,20 +1,7 @@
...
-		),
-		'name' => 'submit',
-		'class' => 'primary',
-		'dropdown' => $dropdown
-	)
-) ?>
+<?php if ( !empty( $approveCheckbox ) ): ?>
+	<label for="wpApprove" class="approve-changes-label">
+		<input id='wpApprove' name='wpApprove' type='checkbox' value='1'/>
+		<?= wfMessage( 'content-review-edit-page-checkbox-label' )->escaped() ?>
```
Comment (kamilkoterba) [inline extensions/wikia/EditPageLayout/templates/EditPageLayout_CodeButtons.php:4]: would be nice to be consistent within file whether to add semicolons at the end of lines or not :)

**Comment by kamilkoterba** [comment]
Great idea on moving MenuButton parameters preparation out of template! :+1:

**Comment by kamilkoterba** [comment]
LGreatTM!

**Comment by kamilkoterba** [comment]
:+1:

**Comment by adamkarminski** [comment]
Tested on a sandbox. Merging.

**Pull request merged**